### DISCUSSION TITLE
SerialUPDI upload fails with huge python error?

### DISCUSSION THREAD

**Discussion by gordonthree** [SpenceKonde/DxCore#317] [Q&A]
Working on a avr128da28 under Arduino 1.8.19 on Fedora, get this error trying to do UPDI upload of a sketch

UPDI does work to upload the bootloader, but my five serial adapters are all busy for diagnostics, so I need to stick with UPDI for uploading.

```
/home/gordon/.arduino15/packages/megaTinyCore/tools/python3/3.7.2-post1/python3 -u /home/gordon/.arduino15/packages/DxCore/hardware/megaavr/1.4.10/tools/prog.py -t uart -u /dev/ttyUSB2 -b 230400 -d avr128da28 --fuses 5:0b11001001 6:0x04 7:0x00 8:0x00 -f/tmp/arduino_build_417907/da28_mands_test.ino.hex -a write -v 
SerialUPDI
UPDI programming for Arduino using a serial adapter
Based on pymcuprog, with significant modifications
By Quentin Bolsee and Spence Konde
Version 1.2.3 - Jan 2022
Using serial port /dev/ttyUSB2 at 230400 baud.
Target: avr128da28
Set fuses: ['5:0b11001001', '6:0x04', '7:0x00', '8:0x00']
Action: write
File: /tmp/arduino_build_417907/da28_mands_test.ino.hex
Traceback (most recent call last):
  File "/home/gordon/.arduino15/packages/DxCore/hardware/megaavr/1.4.10/tools/prog.py", line 285, in <module>
    main()
  File "/home/gordon/.arduino15/packages/DxCore/hardware/megaavr/1.4.10/tools/prog.py", line 127, in main
    setup_logging(user_requested_level=logging_level)
  File "/home/gordon/.arduino15/packages/DxCore/hardware/megaavr/1.4.10/tools/libs/pymcuprog/pymcuprog.py", line 47, in setup_logging
    configfile = yaml.safe_load(file)
  File "/home/gordon/.arduino15/packages/DxCore/hardware/megaavr/1.4.10/tools/libs/yaml/__init__.py", line 94, in safe_load
    return load(stream, SafeLoader)
  File "/home/gordon/.arduino15/packages/DxCore/hardware/megaavr/1.4.10/tools/libs/yaml/__init__.py", line 72, in load
    return loader.get_single_data()
  File "/home/gordon/.arduino15/packages/DxCore/hardware/megaavr/1.4.10/tools/libs/yaml/constructor.py", line 37, in get_single_data
    return self.construct_document(node)
  File "/home/gordon/.arduino15/packages/DxCore/hardware/megaavr/1.4.10/tools/libs/yaml/constructor.py", line 46, in construct_document
    for dummy in generator:
  File "/home/gordon/.arduino15/packages/DxCore/hardware/megaavr/1.4.10/tools/libs/yaml/constructor.py", line 398, in construct_yaml_map
    value = self.construct_mapping(node)
  File "/home/gordon/.arduino15/packages/DxCore/hardware/megaavr/1.4.10/tools/libs/yaml/constructor.py", line 204, in construct_mapping
    return super().construct_mapping(node, deep=deep)
  File "/home/gordon/.arduino15/packages/DxCore/hardware/megaavr/1.4.10/tools/libs/yaml/constructor.py", line 126, in construct_mapping
    if not isinstance(key, collections.Hashable):
AttributeError: module 'collections' has no attribute 'Hashable'
AttributeError: module 'collections' has no attribute 'Hashable'
```

**Comment by gordonthree** [comment]
I am able to program using the standard Microchip programmer:

```
pymcuprog erase  -t uart -d avr128da28 -u /dev/ttyUSB2
Connecting to SerialUPDI
Pinging device...
Ping response: 1E970A
Chip/Bulk erase:
- Memory type eeprom is conditionally erased (depending upon EESAVE fuse setting)
- Memory type flash is always erased
- Memory type lockbits is always erased
Erased.
Done.

pymcuprog write  -t uart -d avr128da28 -u /dev/ttyUSB2 -f /tmp/arduino_build_417907/da28_mands_test.ino.hex
Connecting to SerialUPDI
Pinging device...
Ping response: 1E970A
Writing from hex file...
Writing flash...
Done.
```

**Reply by SpenceKonde** [reply]
> I am able to program using the standard Microchip programmer:
> 
> ```
> pymcuprog erase  -t uart -d avr128da28 -u /dev/ttyUSB2
> Connecting to SerialUPDI
> Pinging device...
> Ping response: 1E970A
> Chip/Bulk erase:
> - Memory type eeprom is conditionally erased (depending upon EESAVE fuse setting)
> - Memory type flash is always erased
> - Memory type lockbits is always erased
> Erased.
> Done.
> 
> pymcuprog write  -t uart -d avr128da28 -u /dev/ttyUSB2 -f /tmp/arduino_build_417907/da28_mands_test.ino.hex
> Connecting to SerialUPDI
> Pinging device...
> Ping response: 1E970A
> Writing from hex file...
> Writing flash...
> Done.
> ```
And then you get up, make a cup of coffee, and by the time you're done finishing the verification.

**Comment by SpenceKonde** [comment]
This is fixed in master. It was an old version of the included YAML library. And this is written in Python not [Ouroboros](https://en.wikipedia.org/wiki/Ouroboros) so they have no prohibition on changes tha aren't backwards compatible.

**Comment by gordonthree** [comment]
Any suggestions how to fix this? Arduino claims there's no update available, is there a way to force it to pull the latest code from GitHub?

**Comment by SpenceKonde** [comment]
You don't want to pull in the code from the github repo, it is completely
hosed atm. I am working full time on fixing this please be patient.

On Mon, Aug 8, 2022 at 10:45 AM GordonThree ***@***.***>
wrote:

> Any suggestions how to fix this? Arduino claims there's no update
> available, is there a way to force it to pull the latest code from GitHub?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SpenceKonde/DxCore/discussions/317#discussioncomment-3350212>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTXEW7L7VWGATIHVEV5JYDVYEMPXANCNFSM552YXQQQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore <https://github.com/SpenceKonde/ATTinyCore>: Arduino support for
all pre-2016 tinyAVR with >2k flash!
megaTinyCore <https://github.com/SpenceKonde/megaTinyCore>: Arduino support
for all post-2016 tinyAVR parts!
DxCore <https://github.com/SpenceKonde/DxCore>: Arduino support for the AVR
Dx-series parts, the latest and greatest from Microchip!
Contact: ***@***.***

**Reply by gordonthree** [reply]
Will do, thank you for all your work on this core!
